lib/bundler/cli/exec.rb

```diff
@@ -81,6 +82,34 @@ def kernel_load(file, *args)
       abort "#{e.class}: #{e.message}\n  #{backtrace.join("\n  ")}"
     end
 
+    # This is a bit esoteric. There's two separate pieces to understand first.
+    #
+    # (1) When `:disable_shared_gems` is true, the `GEM_PATH` gets initialized
+    # to an empty string. This is done to make sure it's expanded to ONLY the
+    # Bundler --path setting, otherwise it expands to include the system path.
+    #
+    # (2) Prior to 1.12, the code here in bundle exec didn't use Kernel.load
+    # and didn't `require "bundler/setup"`, which meant the path to gems was
+    # never involved.
+    #
+    # In 1.12, bundle exec was overhauled for various reasons to use
+    # Kernel.load, and `require "bundler/setup"` is now invoked, which created
+    # a bug. In cases like `--deployment` where `disable_shared_gems` is true,
+    # Bundler couldn't find itself, because Bundler never lives in the
+    # `--path` but only in system gems.
+    #
+    # We fixed this (the bundle exec bug) in 1.13.0 by changing GEM_PATH to be
+    # initialized to nil instead of empty string in all cases. But it created
+    # another bug. We've reverted the change so that GEM_PATH is now back to
+    # being initialized, but still need to override how the GEM_PATH is set
+    # in this special case so that the bundler executable can be found.
+    def configure_gem_path_override
+      if Bundler.settings[:disable_shared_gems]
+        # TODO MODO hate this spooky at a distance global state voodoo
+        ENV["DO_NOT_LIMIT_GEM_PATH"] = "true"
```

Comment:
this `configure_gem_path_override` method is now called _before_ `require "bundler/setup"` ... prior commit had me attempting to undo something that was done inside bundler/setup _after_ it was called ... had to change the approach since it was too late at that point.